### ISSUE TITLE
Scope race conditions

### ISSUE THREAD

**Issue by subzero911** [fluttercommunity/get_it#316]
I pop scope on widget dispose and register a new singleton on a new screen open.
But the new singleton gets into the old scope which would got disposed in a next frame. That's because of screens transitions which has not finished, thus old screen's dispose function has called a bit later than new screen's initState.
I couldn't figure out for a long time why my new singleton disappears!

**Proposal:**

`GetIt.pushNewScope()` and `GetIt.popScope()` shouldn't be imperative.
Force user to register singletons inside of pushNewScope callback
```dart
var profileScope = GetIt.pushNewScope(init: (getIt) { 
  getIt.registerSingleton(EditProfileController());
}) 

GetIt.I.registerSingleton(YetAnotherController()); // does not get into the scope above
```
User should pop scope only by name or by instance:
```dart
  profileScope.pop();
  // or
  GetIt.popScope(profileScope);
  // or
  GetIt.popScope(name: 'profile');
```

also related issues https://github.com/fluttercommunity/get_it/issues/313, https://github.com/fluttercommunity/get_it/issues/303

**Comment by subzero911** [comment]
> isn't there one getIt to much in there?

Oh, that's a typo.
Of course, I meant `GetIt.I.registerSingleton(YetAnotherController());`

**Comment by escamoteur** [comment]
have you checked the get_it_mixin with it's pushScope method?

**Comment by subzero911** [comment]
> have you checked the get_it_mixin with it's pushScope method?

No, I didn't. Could it solve my problem?

**Comment by escamoteur** [comment]
not 100% sure. are you aware, that all the pop functions are async and return a Future?

**Comment by subzero911** [comment]
As far as I can see, pushScope method of `get_it_mixin` is designed to be called in a widget's `build` method only.
But I want to push scopes inside the router config: https://github.com/flutter/flutter/issues/102408#issuecomment-1361256775
```dart
GoRoute(
        path: '/profile',
        builder: (context, state) => const ProfileScreen(),
        onInit: () => GetIt.pushNewScope(init: (getIt) { getIt.registerSingleton(ProfileController());} )
        onDispose: () => GetIt.popScope();
      ), 
```
Also, `get_it_mixin`'s pushScope is still imperative, and it can't be guaranteed that other singletons won't get into this scope right before the widget is disposed.

**Comment by escamoteur** [comment]
yes, you are right. I think the problem is that `PopScope` is async

**Comment by escamoteur** [comment]
OK, sometimes it takes time to fully understand the problem that people have. The obvious solution is https://github.com/fluttercommunity/get_it/pull/292.
So if you only want a scope to live as long as a route or widget you should only use `disposeScope('scopeName') and not mix it with using `popScope`.
Does that make sense? Alternatively, we could think about adding a separate scope list that is not part of the scope stack that has its own `createScope`, `disposeScope` function, but that would make the object lookup slower.

**Comment by escamoteur** [comment]
Included in just pushed V7.5.0

**Comment by subzero911** [comment]
@escamoteur I checked out the new function.
But this new dropScope() method does not protect me from getting other singletons into this scope, unless added yet another scope above.

```dart
GoRoute(
  path: '/profile',
  builder: (context, state) => const ProfileScreen(),
  onInit: () {
    getIt.pushNewScope(scopeName: 'scope1');
    getIt.registerSingleton(Foo());
  }
  onDispose: () {
   getIt.dropScope('scope1')
  }),
GoRoute(
  path: '/details',
  builder: (context, state) => const DetailsScreen(),
  onInit: () => getIt.registerSingleton(Bar());
}
```
On navigating from /profile screen to the /details screen:
Expected behaviour: Bar() is registered in the base scope.
Actual behaviour: Bar() is registered in the 'scope1', then 'scope1' gots disposed.
Your fix does nothing about it.
